src/BenchmarksApps/SslStream/Server/Program.cs

```diff
@@ -0,0 +1,177 @@
+
+        CancellationTokenSource ctrlC = new CancellationTokenSource();
+        Console.CancelKeyPress += (s, e) =>
+        {
+            Log("Shutting down...");
+            e.Cancel = true;
+            ctrlC.Cancel();
+        };
+
+        try
+        {
+            while (!ctrlC.IsCancellationRequested)
+            {
+                var clientSock = await sock.AcceptAsync(ctrlC.Token).ConfigureAwait(false);
+                _ = Task.Run(() => ProcessClient(clientSock, options, sslOptions, ctrlC.Token));
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Expected, just return
+        }
+
+        Log("Exitting...");
+    }
+
+    static async Task ReadWriteScenario(SslStream stream, ServerOptions options, CancellationToken cancellationToken)
+    {
+        static async Task WritingTask(SslStream stream, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (bufferSize == 0)
+            {
+                return;
+            }
+
+            var sendBuffer = new byte[bufferSize];
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    await stream.WriteAsync(sendBuffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected, just return
+            }
+        }
+
+        byte[] recvBuffer = new byte[options.ReceiveBufferSize];
+
+        CancellationTokenSource cts = new CancellationTokenSource();
+        using var reg = cancellationToken.Register(() => cts.Cancel());
+
+        var writeTask = Task.Run(() => WritingTask(stream, options.SendBufferSize, cts.Token));
+
+        var recvBufer = new byte[options.ReceiveBufferSize];
+
+        while (true)
+        {
+            if (await stream.ReadAsync(recvBufer, cts.Token).ConfigureAwait(false) == 0)
+            {
+                cts.Cancel();
+                break;
+            }
+        }
+
+        await writeTask.ConfigureAwait(false);
+    }
+
+    static async Task ProcessClient(Socket socket, ServerOptions options, SslServerAuthenticationOptions sslOptions, CancellationToken cancellationToken)
+    {
+        try
+        {
+            using var stream = await EstablishSslStreamAsync(socket, sslOptions, cancellationToken).ConfigureAwait(false);
+
+            if (stream.NegotiatedApplicationProtocol.Protocol.Span.SequenceEqual("Handshake"u8))
+            {
+                // done everything
+                return;
+            }
+            if (stream.NegotiatedApplicationProtocol.Protocol.Span.SequenceEqual("ReadWrite"u8))
+            {
+                await ReadWriteScenario(stream, options, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                Log($"Negotiated unknown protocol: {stream.NegotiatedApplicationProtocol}");
```

Comment:
Expanding the comment on the socket exception -- it made me think about unhandled exceptions -- I think we should treat some errors as "unrecoverable", meaning the benchmark itself must be set up incorrectly for us to get such an error. So we need to log, but should not prevent the exception from bubbling up. Then the benchmark will be aborted, and the error together with the logs will be reported in crank output.

E.g. if we suddenly negotiated an unknown protocol, it looks like something went wrong and we should not proceed further :)

Reply:
We can't easilly bubble it up out of Main because we hand off incomming connections to a task on threadpool, but I can add a metric for errors.